### ISSUE TITLE
Try Flamingo approach to solar PV forecasting

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/power_perceiver#155]
* Train a ResNet auto-encoder to predict the next frame of satellite imagery (or an arbitrary frame, and give it the forecast step?)
* Freeze (or don't freeze?) the ResNet and feed its "code layer" into a Perceiver (see issue #154). 
* Perhaps we directly query the Perceiver to ask for the PV power predictions.
* Or, more like Flamingo, feed the output of the Perceiver into a Transformer Decoder which forecasts PV (perhaps like the Temporal Fusion Transformer)

Could train the ResNet on "small" patches of imagery (256x256) and then use several ResNets to "see" the entire country? (Then we'd have to use our own spatial encodings. But that's fine)

**Comment by JackKelly** [comment]
Hi @bndxn! The main motivation for predicting satellite imagery was to allow us to pre-train the "satellite predictor" on any rectangle of satellite imagery, even locations with no PV systems (eg over the ocean!). This way, we hoped to get the model to learn cloud dynamics from as much data as possible. The idea is that predicting the movement of clouds is hard, so we want to train the model on as much data as possible. 

I still very much believe in the general principle that we want to pre-train part of the ML model on as much satellite imagery as possible! But I now suspect that predicting pixel-by-pixel images is just too onorous. So it might be better to learn an encoder whose latent representation is maximally informative of the latent representation of future satellite imagery

**Comment by bndxn** [comment]
Thanks for getting back to me. Cool, that makes a lot of sense. I'm guessing when there's no PV data, the labels/target is a prediction of the same rectangle, is that right? 

I can see how this approach would help generate good predictions of cloud patterns, and I can see how big cloudy patches would have a direct link to PV. Are there some nuances in clouds that this approach would help for, but that don't matter very much from a PV perspective? For example, maybe the satellite imagery gets very good at predicting lots of fine thin cloud vs thin ripply cloud (I am not a cloud expert!) but actually from a PV yield view these two are the same - does that make sense?

**Comment by JackKelly** [comment]
That's a good example!

My hunch is that the problem is that it's _really_ hard to predict clouds on a pixel-by-pixel basis. So we're perhaps hurting the model by asking it to do something that's (almost) impossible! So we might be better off predicting high-level, abstract features of the satellite imagery, so we can still learn how clouds evolve from all the available satellite data.

**Comment by JackKelly** [comment]
PyTorch implementation of Contrastive Predictive Coding: https://github.com/rschwarz15/CPCV2-PyTorch

(Hat tip to @jacobbieker! Thanks Jacob!)

**Comment by JackKelly** [comment]
(Actually, my new favourite idea is described in openclimatefix/psuedo-pv-labeller#1! I plan to try openclimatefix/psuedo-pv-labeller#1 before I try contrastive learning. openclimatefix/psuedo-pv-labeller#1 feels like it has many of the same advantages, without some of the complexity of contrastive learning)

**Comment by bndxn** [comment]
Cool, that sounds really interesting. At the moment I'm working with a [subset](https://github.com/bndxn/dissertation/blob/main/Input_5_Merging_inputs_and_inspecting.ipynb) of the data and doing a [ConvLSTM](https://github.com/bndxn/dissertation/blob/main/Model_12_ConvLSTM_single_input%2C_1hr_back%2C_2hr_forward.ipynb) to predict PV directly going forwards. 

My plan is to compare this with a simplified version of OCF's CNN in production (maybe only taking sat images and the solar altitude as inputs). The idea is to compare explicitly how important the explicit representation of time series in ConLSTMs is, compared to a 3D CNN. 

If I can wrangle that in the next few weeks, then maybe I'll include a comparison with an attention model. If I finish all of that before September, then yeah sure! This will all provide helpful context anyway.

**Comment by JackKelly** [comment]
Sounds awesome! Please do let us know how it goes! Very exciting stuff!
